### PR TITLE
Aggregations Refactor: Refactor Scripted Metric Aggregation

### PR DESCRIPTION
**Pull request by colings86**

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/ScriptedMetricAggregator.java [hunk 1]
```diff
@@ -188,6 +222,97 @@ private static Script deepCopyScript(Script script, SearchContext context) {
...
+            out.writeBoolean(hasReduceScript);
+            if (hasReduceScript) {
+                reduceScript.writeTo(out);
+            }
+            boolean hasParams = params != null;
+            out.writeBoolean(hasParams);
+            if (hasParams) {
+                out.writeMap(params);
+            }
+        }
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/ScriptedMetricAggregator.java]: should we use writeOptionalStreamable/readOptionalStreamable for serialization instead?

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/ScriptedMetricTests.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        if (randomBoolean()) {
+        factory.initScript(randomScript("initScript"));
+        }
+        factory.mapScript(randomScript("mapScript"));
+        if (randomBoolean()) {
+        factory.combineScript(randomScript("combineScript"));
+        }
+        if (randomBoolean()) {
+        factory.reduceScript(randomScript("reduceScript"));
+        }
```
Comment (jpountz) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/ScriptedMetricTests.java]: the above if statements have indentation issues?

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/ScriptedMetricTests.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        return factory;
+    }
+
+    private Script randomScript(String script) {
+        int initBranch = randomInt(3);
+        switch (initBranch) {
+        case 0:
+            return new Script(script);
+        case 1:
+        default:
```
Comment (jpountz) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/ScriptedMetricTests.java]: let's remove "case 1:" unless maybe you actually wanted to put something there but forgot to?

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/ScriptedMetricAggregator.java (hunk 1 above)
Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/ScriptedMetricAggregator.java]: implemented

File: core/src/main/java/org/elasticsearch/script/Script.java [hunk 4]
```diff
@@ -34,12 +34,20 @@
...
 import java.io.IOException;
 import java.util.Map;
+import java.util.function.Supplier;
 
 /**
  * Script holds all the parameters necessary to compile or find in cache and then execute a script.
  */
 public class Script implements ToXContent, Streamable {
 
+    public static final Supplier<Script> SUPPLIER = new Supplier<Script>() {
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/script/Script.java]: can you add javadocs that this is for usage in conjunction with StreamInput.readStreamable?

**Comment by jpountz** [comment]
LGTM

**Pull request merged**